### PR TITLE
[logreader] remove embedded commons-io library

### LINKED ISSUE #5442: Builds fail in logreader
Since today all builds fail while compiling logreader. This is happening for Travis, Jenkins and in my local builds. 

```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project org.openhab.binding.logreader: Compilation failure
[ERROR] C:\Users\jnk\openhab2-master\git\openhab2-addons\bundles\org.openhab.binding.logreader\src\main\java\org\openhab\binding\logreader\internal\filereader\FileTailer.java:[64,1989] The constructor Tailer(File, TailerListener, long, boolean, boolean, boolean) is undefined
```

This constructor is the reason why the commons.io-library is included in the binding since it is not merged upstream and Eclipse is able to resolve it.

### PR DESCRIPTION
**Pull request by J-N-K**
Fixes #5442 

adds modified class files instead of embedding custom compiled library

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by J-N-K** [comment]
I had to suppress the header and author checks. Is that right? Adding unchanged source files and changing the header doesn't sound right to me.

**Comment by Hilbrand** [comment]
This is the wrong solution. It's just patching the problem and adding technical depth. The right solution would be to update the commons io library in core to a newer version.

**Comment by J-N-K** [comment]
Not possible, the required update is not available. 

Latest release is 2.6, 2.7 is unreleased and the fix is not even merged for 2.7.

**Comment by Hilbrand** [comment]
Ok should have read your issue first :smile: Is it still an idea to do add the dependency and only add the modified file?

**Comment by J-N-K** [comment]
I'm not sure what you are suggesting. 

There are only changes to `Tailer.java`. Unfortunately we can't use the inherit/overwrite solution (which would make it mor clear IMO), because the method we would need to override is public but needs access to private members. Integrating the class needs a changed package which in turn required the other two classes.

The problem is that the compiler is not picking up the embedded library if the same package is already available. This will not change if we embed only these files as external library. I tried to fix that but did not succeed.

And I don't think that upgrading commons-io to an unreleased version for all bundles is a good idea. 

This solution at least makes our build working again. The other solution would be to revert #5383.

File: tools/checkstyle/suppressions.xml [hunk 1]
```diff
@@ -19,4 +19,6 @@
     <suppress files=".+org.openhab.binding.wifiled.handler.FadingWiFiLEDDriver.java" checks="AvoidScheduleAtFixedRateCheck"/>
     <suppress files=".+[\\/]pom\.xml" checks="OnlyTabIndentationCheck|OnlyTabIndentationInXmlFilesCheck"/>
     <suppress files=".+org.openhab.binding.yeelight.+" checks="OutsideOfLibExternalLibrariesCheck" />
+	    <!-- suppress header checks for imported and patched apache commons-io files in logreader binding -->
```
Comment (Hilbrand) [inline tools/checkstyle/suppressions.xml]: tab?

**Comment by Hilbrand** [comment]
If @kaikreuzer agrees with adding source files with a different license than I have no objection. It's just we always tell not sources files with other licenses should be added. So this breaks that policy. I would not have an objection or revert the migration so we can find a better solution. Redoing the migration later on is very simple.

**Comment by davidgraeff** [comment]
It is the Apache 2 license, the partner license to the Eclipse license. Apache is allowed in openhab core. I also wonder why only Kai can decided that?

**Comment by kaikreuzer** [comment]
> It's just we always tell sources files with other licenses should be not be added. So this breaks that policy.

Correct. We must not end up having different files under different licenses, this would be a mess. We have the NOTICE files stating that it is EPLv2 and this is what contributors automatically assume as well, when doing contributions. Furthermore, our automatic header update (through `mvn license:format` will mess up all headers in such situations...).

Despite this, we already had [similar situations in the past](https://github.com/openhab/openhab2-addons/pull/3582#discussion_r190253129) where we had to add foreign source code. The solution we chose was to place it in a separate source folder `src/3rdparty/java` and specifically mentioned this code and its license in the bundles NOTICE file (just like we mention binary 3rdparty libs in there as well). I'd think, the same could be applied here, wdyt?

**Comment by J-N-K** [comment]
Unfortunately src/3rdparty is not found at build time. And 3rdparty is not a valid package name. I chose `src/main/java/org/openhab/binding/logreader/internal/thirdparty` and made a note in the NOTICE file.

**Comment by kaikreuzer** [comment]
> Unfortunately src/3rdparty is not found at build time

You will have to add `src/3rdparty/java` as an additional source path in the Maven pom (by default, Maven only looks for `src/main/java`). See e.g. https://www.baeldung.com/maven-project-multiple-src-directories.

> And 3rdparty is not a valid package name.

`src/3rdparty/java` is the source folder name, NOT a Java package name. I'd actually suggest to keep the original Java package name (i.e. `org.apache....`), so that it is clear that this is code from 3rd party, which we do not maintain in our repo, but which has only been copied there.

**Comment by J-N-K** [comment]
Thanks. Keeping the package names is unfortunately not possible since this conflicts with org.apache.commons.io.input:2.2 which is already available on build path. If I force the orginal name, the wrong version of the class is pickled up (and we are where we started).

**Comment by kaikreuzer** [comment]
Thanks for the update!
What about my comment: "I'd actually suggest to keep the original Java package name (i.e. org.apache....)"? Is there any good reason why the files should be under org.openhab namespace?

**Comment by kaikreuzer** [comment]
> If I force the orginal name, the wrong version of the class is pickled up (and we are where we started).

This should be only a matter of sorting the classpath - it should use your sources first before going to the dependencies.
Where did you experience problems? In the IDE or with the Maven build?

**Comment by J-N-K** [comment]
Strange. It worked after I tried that again.

**Comment by martinvw** [comment]
Let's wait for the build :-)

**Pull request merged**